### DISCUSSION TITLE
Working with custom metric for XGBoost

### DISCUSSION THREAD

**Discussion by DavidHarar** [optuna/optuna#4576] [Q&A]
### Don't use GitHub Issues to ask support questions.

I am trying to add a custom metric to my study, and optimize the model by it. I searched for a working code without any success. I found an example for lgbm [here](https://stackoverflow.com/questions/74303961/custom-eval-metric-using-early-stopping-in-lgbm-sklearn-api-and-optuna) and combined it with the parametrization from [here](https://www.kaggle.com/code/kst6690/make-your-xgboost-model-awesome-with-optuna), where they added the custom metric in the following fashion 
  
```
param_grid = {'max_depth':..,
                         ...,
                         'eval_metric':trial.suggest_categorical('eval_metric', metric_list),
                         ...}
```
  
When running the study I get results on the training set but none on the validation set, and after one trial I get an error. I didn't want to add it as an issue because it is probably me who doesn't understand how to use Optuna properly. I added an example of my code below and the results I get.  

```
def custom_metric(y_true, y_pred):
    kpi_value = some_function(y_true, y_pred)
    return kpi_value

def objective(trial):

    metric_list = [custom_metric]
    param_grid = {
        "verbosity":0,
        "objective": "binary:logistic",
        "booster": "gbtree",
        'eval_metric':trial.suggest_categorical('eval_metric', metric_list),
        "lambda": trial.suggest_loguniform("lambda", 1e-8, 1.0),
        "alpha": trial.suggest_loguniform("alpha", 1e-8, 1.0),
        "max_depth": trial.suggest_int("max_depth", 9, 25),
        "eta": trial.suggest_loguniform("eta", 1e-8, 1.0),
        "gamma": trial.suggest_loguniform("gamma", 1e-8, 1.0),
        "early_stopping_rounds":15
    }

    train_x, valid_x, train_y, valid_y = train_test_split(X_train, y_train, test_size=0.25)
    model = xgb.XGBClassifier(
            **param_grid,
            n_jobs=-1
        )

    model.fit( 
        train_x,
        train_y,
        eval_set=[(valid_x, valid_y)],
        # callbacks=[ # NOTE: when adding the pruning I get an error without message.
        #     optuna.integration.XGBoostPruningCallback(trial, custom_metric),  # THIS HAS CHANGED (REF QUESTION 2)!
        # ]
    )
    preds = model.predict_proba(valid_y)[:,1]
    kpi_value_ = custom_metric(valid_y, preds)
    return kpi_value_
```

The results I get are the following:
  
```
[0]	validation_0-logloss:0.68601	validation_0-custom_metric:0.00000
[1]	validation_0-logloss:0.68010	validation_0-custom_metric:0.00000
[2]	validation_0-logloss:0.67493	validation_0-custom_metric:0.00158
[3]	validation_0-logloss:0.67030	validation_0-custom_metric:0.00158
[4]	validation_0-logloss:0.66551	validation_0-custom_metric:0.00000
[5]	validation_0-logloss:0.66139	validation_0-custom_metric:0.00158
[6]	validation_0-logloss:0.65725	validation_0-custom_metric:0.00158
[7]	validation_0-logloss:0.65356	validation_0-custom_metric:0.00000
[8]	validation_0-logloss:0.65066	validation_0-custom_metric:0.00000
[9]	validation_0-logloss:0.64788	validation_0-custom_metric:0.00158
[10]	validation_0-logloss:0.64483	validation_0-custom_metric:0.00088
[11]	validation_0-logloss:0.64222	validation_0-custom_metric:0.00000
[12]	validation_0-logloss:0.63989	validation_0-custom_metric:0.00000
[13]	validation_0-logloss:0.63851	validation_0-custom_metric:0.00000
[14]	validation_0-logloss:0.63647	validation_0-custom_metric:0.00000
[15]	validation_0-logloss:0.63461	validation_0-custom_metric:0.00000
[W 2023-04-03 16:25:04,896] Trial 0 failed with parameters: {'eval_metric': <function custom_metric at 0x7fd5e80ed820>, 'lambda': 0.01383704932819595, 'alpha': 1.8033248603918073e-08, 'max_depth': 17, 'eta': 0.030946019456836216, 'gamma': 1.1487064299469371e-06} because of the following error: ValueError('training data did not have the following fields: var1, var2, var32).
...
```
  
And I also know for sure that `var1, var2, var32` are in both datasets.
